lib/bgpstream_bgpdump.c

```diff
@@ -0,0 +1,234 @@
+      buf_p++;                                                                 \
+    }                                                                          \
+  } while (0)
+
+char *bgpstream_record_elem_bgpdump_snprintf(char *buf, size_t len,
+                                     bgpstream_record_t *record,
+                                     bgpstream_elem_t *elem)
+{
+  assert(record);
+  assert(elem);
+
+  size_t written = 0; /* < how many bytes we wanted to write */
+  ssize_t c = 0;      /* < how many chars were written */
+  char *buf_p = buf;
+
+  /* Record type */
+  switch (elem->type) {
+  case BGPSTREAM_ELEM_TYPE_RIB:
+    c = snprintf(buf_p, B_REMAIN, "TABLE_DUMP2|%u", record->time_sec);
+    break;
+  case BGPSTREAM_ELEM_TYPE_ANNOUNCEMENT:
+  case BGPSTREAM_ELEM_TYPE_WITHDRAWAL:
+  case BGPSTREAM_ELEM_TYPE_PEERSTATE:
+    c = snprintf(buf_p, B_REMAIN, "BGP4MP|%u", record->time_sec);
+    break;
+  default:
+    break;
+  }
+  written += c;
+  buf_p += c;
+  ADD_PIPE;
+
+  switch(elem->type){
+  case BGPSTREAM_ELEM_TYPE_RIB:
+    c = snprintf(buf_p, B_REMAIN, "B");
+    break;
+  case BGPSTREAM_ELEM_TYPE_ANNOUNCEMENT:
+    c = snprintf(buf_p, B_REMAIN, "A");
+    break;
+  case BGPSTREAM_ELEM_TYPE_WITHDRAWAL:
+    c = snprintf(buf_p, B_REMAIN, "W");
+    break;
+  case BGPSTREAM_ELEM_TYPE_PEERSTATE:
+    c = snprintf(buf_p, B_REMAIN, "STATE");
+    break;
+  default:
+    break;
+  }
+
+  written += c;
+  buf_p += c;
+  ADD_PIPE;
+
+  /* PEER IP */
+  if (bgpstream_addr_ntop(buf_p, B_REMAIN, &elem->peer_ip) == NULL) {
+    bgpstream_log(BGPSTREAM_LOG_ERR, "Malformed peer address");
+    return NULL;
+  }
+  SEEK_STR_END;
+  ADD_PIPE;
+
+  /* PEER ASN */
+  c = snprintf(buf_p, B_REMAIN, "%" PRIu32, elem->peer_asn);
+  written += c;
+  buf_p += c;
+  ADD_PIPE;
+
+  switch(elem->type){
+  case BGPSTREAM_ELEM_TYPE_RIB:
+  case BGPSTREAM_ELEM_TYPE_ANNOUNCEMENT:
+    /* PREFIX */
+    if (bgpstream_pfx_snprintf(buf_p, B_REMAIN,
+                               (bgpstream_pfx_t *)&(elem->prefix)) == NULL) {
+      bgpstream_log(BGPSTREAM_LOG_ERR, "Malformed prefix (R/A)");
+      return NULL;
+    }
+    SEEK_STR_END;
+    ADD_PIPE;
+
+    /* AS PATH */
+    c = bgpstream_as_path_snprintf(buf_p, B_REMAIN, elem->as_path);
+    written += c;
+    buf_p += c;
+    if (B_FULL)
+      return NULL;
+    ADD_PIPE;
+
+    /* SOURCE (IGP) */
+    switch(elem->origin){
+    case BGPSTREAM_ELEM_BGP_UPDATE_ORIGIN_IGP:
+      c = snprintf(buf_p, B_REMAIN, "IGP");
+      break;
+    case BGPSTREAM_ELEM_BGP_UPDATE_ORIGIN_EGP:
+      c = snprintf(buf_p, B_REMAIN, "EGP");
+      break;
+    case BGPSTREAM_ELEM_BGP_UPDATE_ORIGIN_INCOMPLETE:
+      c = snprintf(buf_p, B_REMAIN, "INCOMPLETE");
+      break;
+    default:
+      break;
+    }
+    written += c;
+    buf_p += c;
+    ADD_PIPE;
+
+    /* NEXT HOP */
+    if (bgpstream_addr_ntop(buf_p, B_REMAIN, &elem->nexthop) != NULL) {
```

Comment:
why this way around rather than the other way used above?
```
  if (bgpstream_addr_ntop(buf_p, B_REMAIN, &elem->peer_ip) == NULL) {
    bgpstream_log(BGPSTREAM_LOG_ERR, "Malformed peer address");
    return NULL;
  }
```